Add a page about the reconciler pattern in a clustered DBMS (#866)
neo4j/docs-operations: modules/ROOT/pages/clustering/clustering-advanced/reconciler.adoc

[role=enterprise-edition]
[[cluster-reconciler]]
= Reconciler
:description: This section describes how changes to the DBMS are processed by each server.

> [[cluster-reconciler-introduction]]

Reviewer: Are all the section labels needed? Do you want specific IDs for each of them? Applies to later sections as well.

Author: Ah, I thought I was supposed to label all sections. I don't think deep linking to a section on a page this small is necessarily important, happy to remove them.

== Introduction

> Administrative operations in Neo4j, such as creating a database, do not happen synchronously.

Reviewer: ```suggestion
In Neo4j, administrative operations such as database creation do not happen synchronously.
```

> Instead we have a model where the changes are written to the `system` database, and record the desired state of the DBMS.

Reviewer: ```suggestion
Instead, the changes are first written to the `system` database, recording the desired state of the DBMS.
```

> Each server has an internal component called the reconciler, which observes the desired state and makes changes to the local server to match that state.

Reviewer: ```suggestion
Each server has a _reconciler_, an internal component which observes the desired state and makes changes to the local server to match that state.
```

> For example, the `system` database may record that it wants the database `bar` running on `server-1`.

Reviewer: ```suggestion
For example, the `system` database may record that the database `bar` is expected to be running on `server-1`.
```

> When the reconciler on `server-1` sees that, it will start `bar`.

Reviewer: ```suggestion
When the reconciler on `server-1` becomes aware of that, it starts `bar`.
```

> Executing an operation and getting back a successful response means that the desire is safely committed to the `system` database, and will be processed by each member of the cluster at some point, assuming the server is healthy enough.

Reviewer: ```suggestion
Executing an operation and getting back a successful response means that the desired state is safely committed to the `system` database, and that it will be processed by each member of the cluster at some point (assuming the server is healthy enough).
```
`desire` as a verb or noun is _very_ strong https://dictionary.cambridge.org/dictionary/english/desire
The adjective is okay, but I recommend reviewing the whole text and checking if another adjective is possible. For example, how about _intended state_?

Author: We use `requestedHosting` in `SHOW SERVERS`, _intended_ is fine by me but _requested_ is an existing term, so probably preferable?

Reviewer: Yeah, _requested_ is also fine!

[[cluster-reconciler-async]]

> == Asynchronicity

Reviewer: I recommend taking away both this and the next section and just incorporating these two paragraphs into the main text. Separate sections for 2 sentences each is quite short.

> Servers can observe the same operation at different times, since the `system` database is just another Raft group, where followers and replicas can lag behind the leader in some situations.

Reviewer: ```suggestion
Servers can become aware the same operation at different times, since the `system` database is just another Raft group, where followers and replicas can lag behind the leader in some situations.
```

> Each server should eventually see the new state though, and take action on it.

Reviewer: This remark leaves readers with doubt. _Are there circumstances where the new state is not propagated? Do I need to be careful and account for it? Are there known issues on this?_
Can we make it affirmative, as in 

> At all rates/Sooner or later, the new state gets eventually propagated to all servers.

Author: The doubt is sort-of deliberate. There _are_ circumstances where a server will not see the results, though they are examples of bad problems. The server being permanently offline, long network partitions etc.

I could add a different qualifier, like:
> Eventually, the new state gets propagated to all healthy servers.

Reviewer: > Eventually, the new state gets propagated to all healthy servers.

This sounds great.

[[cluster-reconciler-wait]]
== Waiting

> If you want to be sure that each server has processed an administrative operation before you run the next action, you can use the `WAIT` keyword.

Reviewer: ```suggestion
If you want to be sure that each server has processed an administrative operation before running the next action, you can use the link:{neo4j-docs-base-uri}/cypher-manual/current/administration/databases/#administration-wait-nowait[`WAIT`] keyword.
```

`WAIT` makes the statement not return until all servers have seen the transaction, and their reconciler has finished processing it.

[NOTE]
====

> A statements with `WAIT` returning does not mean that your operation is necessarily successful everywhere.

Reviewer: ```suggestion
When a statements with `WAIT` returns, it does not mean that the operation was necessarily successful everywhere.
```

`WAIT` just ensures the operation has been processed.

> A request to start a database has still been processed if the reconciler tried to start it and the database failed and went into the dirty state.

Reviewer: ```suggestion
For example, a request to start a database is considered _processed_ if the reconciler tried to start it, but the database failed and went into the dirty state.
```

====

> === Example using `WAIT`

Reviewer: ```suggestion
.Create database with `WAIT`
```

> [source, cypher, role="noplay"]

Reviewer: ```suggestion
[source, cypher]
```

The play button has been taken away some months back so `noplay` has no effect anymore.

----
CREATE DATABASE foo WAIT
----

[[cluster-reconciler-errors]]
== Errors

An operation might only succeed on some servers.

> For instance some servers might be offline when you create a database, or a disk error might cause a server to fail to start a database.

Reviewer: ```suggestion
For instance, some servers might be offline when a database is created, or a disk error might cause a server to fail to start a database.
```

> In this situation the operation has not 'failed' as a whole, since you may even have a running database, just with less fault tolerance than you desired.

Reviewer: ```suggestion
In this situation, the operation has not _failed_ as a whole, since the database may even be running (although with less fault tolerance than you intended).
```

> Some failures can just be re-attempted, for example `START DATABASE` can be run if not all members started successfully, and you think the reason has been resolved.

Reviewer: ```suggestion
Some failed operations can just be re-tried: for example, `START DATABASE` can be re-run if not all members started successfully, and you think the cause of that has been resolved.
```

> Other failures will need just the underlying problem to be resolved and the reconciler will make the change, because it is still trying to achieve the desired state.

Reviewer: ```suggestion
Other failures only require the underlying problem to be resolved. 
When the issue is fixed, the reconciler will make the intended change, because it is still trying to achieve the desired state.
```

> For example if a new server fails during a `DEALLOCATE DATABASES FORM SERVER`, meaning a database cannot be safely moved, fixing the new server should be enough to resolve the problem, as the new server will start its copy of the database, allowing the deallocating server to shut down its copy.

Reviewer: ```suggestion
For example, if a new server fails during a `DEALLOCATE DATABASES FROM SERVER` (meaning a database cannot be safely moved), fixing the new server should be enough to resolve the problem, as the new server will start its copy of the database as soon as the target is ready, allowing the deallocating server to shut down its copy.
```

[[cluster-reconciler-reason]]

> == Why use the reconciler pattern

Reviewer: Do we need to justify to customers _why_ we use this design? Do we really need this section? Also, I believe any tech person realizes that a distributed architecture is more robust than a centralized one :)
But if we really do, I'd recommend pruning the section a bit and merging it with the introduction (i.e. taking away the section header).
